File: lib/ex_gix/object_detached.ex
defmodule ExGix.ObjectDetached do
  @moduledoc """
  A detached, self-contained object, without access to its source repository.
  """

  @type t :: reference()
  @type kind :: :tree | :blob | :commit | :tag

  @doc """
  The id of the object.
  """
  @spec id(t()) :: ExGix.ObjectId.t()
  def id(object) when is_reference(object) do
    ExGix.Native.object_detached_id(object)
  end

  @doc """
  The kind of the object.
  """
  @spec kind(t()) :: kind()
  def kind(object) when is_reference(object) do
    ExGix.Native.object_detached_kind(object)
  end

  @doc """
  The fully decoded object data.
  """
  @spec data(t()) :: binary()
  def data(object) when is_reference(object) do
    ExGix.Native.object_detached_data(object)
  end
end

File: native/ex_gix/src/object_detached.rs
use crate::object_id::ObjectIdResource;
use rustler::ResourceArc;

pub struct ObjectDetachedResource(pub gix::ObjectDetached);

rustler::atoms! {
    tree,
    blob,
    commit,
    tag,
}

pub fn kind_to_atom(kind: gix::object::Kind) -> rustler::Atom {
    match kind {
        gix::object::Kind::Tree => tree(),
        gix::object::Kind::Blob => blob(),
        gix::object::Kind::Commit => commit(),
        gix::object::Kind::Tag => tag(),
    }
}

#[rustler::nif]
pub fn object_detached_id(
    obj: ResourceArc<ObjectDetachedResource>,
) -> ResourceArc<ObjectIdResource> {
    ResourceArc::new(ObjectIdResource(obj.0.id))
}

#[rustler::nif]
pub fn object_detached_kind(obj: ResourceArc<ObjectDetachedResource>) -> rustler::Atom {
    kind_to_atom(obj.0.kind)
}

#[rustler::nif]
pub fn object_detached_data<'a>(
    env: rustler::Env<'a>,
    obj: ResourceArc<ObjectDetachedResource>,
) -> rustler::Binary<'a> {
    let slice = obj.0.data.as_slice();
    let mut erl_bin = rustler::OwnedBinary::new(slice.len()).unwrap();
    erl_bin.as_mut_slice().copy_from_slice(slice);
    erl_bin.release(env)
}
